src/mem/cache/prefetch/bop.cc

```diff
@@ -201,27 +207,34 @@ BOP::bestOffsetLearning(Addr x)
         }
     }
 
+    // Move the offset iterator forward to prepare for the next time
     offsetsListIterator++;
 
     // All the offsets in the list were visited meaning that a learning
     // phase finished. Check if
     if (offsetsListIterator == offsetsList.end()) {
         offsetsListIterator = offsetsList.begin();
         round++;
+    }
 
-        // Check if the best offset must be updated if:
-        // (1) One of the scores equals SCORE_MAX
-        // (2) The number of rounds equals ROUND_MAX
-        if ((bestScore >= scoreMax) || (round == roundMax)) {
-            bestOffset = phaseBestOffset;
+    // Check if its time to re-calculate the best offset
+    if ((bestScore >= scoreMax) || (round >= roundMax))
+    {
             round = 0;
+            if (bestScore > badScore)
+            {
+                bestOffset = phaseBestOffset;
+                round = 0;
```

Comment:
We already assign 0 to round, bestScore and phaseBestOffset, and call resetScores outside this conditional block, so something is looking fishy here. Is it a bug or is it duplicated code?

Reply:
Fixed in the latest commit

Reply:
It does not seem to have changed in the latest commit

Reply:
Fixed now. It was indeed duplicate code. Thanks for pointing it out. I've removed the redundant code